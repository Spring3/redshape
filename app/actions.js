export const LOGIN = 'LOGIN';
const login = data => ({ type: LOGIN, data });

export const LOGOUT = 'LOGOUT';
const logout = () => ({ type: LOGOUT });

export const UPDATE_SETTINGS = 'UPDATE_SETTINGS';
const updateSettings = data => ({ type: UPDATE_SETTINGS, data });

export default {
  user: {
    login,
    logout
  },
  settings: {
    update: updateSettings
  }
};
