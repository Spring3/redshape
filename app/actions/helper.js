import * as axios from '../../modules/request';

function IssueFilter () {
  const filter = {};

  this.issue = (ids) => {
    filter.issue_id = Array.isArray(ids) ? ids.join(',') : ids;
    return this;
  };

  this.project = (id) => {
    filter.project_id = id;
    return this;
  };

  this.subproject = (id) => {
    filter.subprojectId = id;
    return this;
  };

  this.tracker = (id) => {
    filter.tracker_id = id;
    return this;
  };

  this.status = ({ open, closed }) => {
    if (open && closed) {
      filter.status_id = '*';
    } else {
      filter.status_id = open ? 'open' : 'closed';
    }
    return this;
  };

  this.assignee = (id) => {
    filter.assigned_to_id = id;
    return this;
  };

  this.author = (id) => {
    filter.author_id = id;
    return this;
  };

  this.title = (str) => {
    if (str) {
      filter.subject = encodeURIComponent(`~${str.trim()}`);
    }
    return this;
  };

  this.due = (date) => {
    filter.due_date = date.toISOString();
    return this;
  };

  this.createdAt = (date) => {
    filter.created_on = date.toISOString();
    return this;
  };

  this.createdBetween = (startDate, endDate) => {
    filter.created_on = encodeURIComponent(`><${startDate.toISOString()}|${endDate.toISOString()}`);
    return this;
  };

  this.priority = (id) => {
    filter.priority_id = id;
    return this;
  };

  this.sort = (by, direction) => {
    if (by && direction) {
      const column = by.indexOf('.') !== -1
        ? by.substring(0, by.indexOf('.'))
        : by;
      filter.sort = `${column}:${direction}`;
    }
    return this;
  };

  this.build = () => ({ ...filter });
}

const handleReject = (e) => {
  console.log('EEEE', e);
  // if this request was not cancelled
  if (!axios.default.isCancel(e)) {
    return Promise.reject(new Error(`Error ${e.status} (${e.statusText || e.message})`));
  }
  console.log('Request was cancelled');
  return undefined;
};

const request = ({
  url,
  method = 'GET',
  data,
  headers,
  query,
  id
}) => {
  const requestConfig = {
    url,
    method,
    headers: axios.makeCancellable(headers || {}, id)
  };

  if (!['GET', 'DELETE'].includes(method)) {
    requestConfig.data = typeof data === 'string' ? data : JSON.stringify(data);
  }

  if (query) {
    requestConfig.params = query;
  }

  if (!axios.getInstance()) {
    return Promise.reject(new Error('401 - Unauthorized'));
  }

  return axios.getInstance().request(requestConfig)
    .then((res) => {
      delete axios.pendingRequests[id];
      return ({ data: res.data });
    })
    .catch((error) => {
      delete axios.pendingRequests[id];
      return handleReject(error);
    });
};

const login = ({
  redmineEndpoint,
  url,
  headers
}) => {
  const defaultConfig = axios.getDefaultConfig();
  return axios.default.request({
    baseURL: redmineEndpoint,
    timeout: defaultConfig.timeout,
    headers: {
      ...defaultConfig.headers,
      ...(headers || {})
    },
    url,
    method: 'GET'
  }).then((res) => {
    axios.reset();
    const { api_key } = res.data.user || {};
    if (api_key) {
      axios.initialize(redmineEndpoint, api_key);
    }
    return { data: res.data };
  });
};

const notify = {
  start: (type, id) => ({ type, status: 'START', id }),
  paginate: (type, data, id) => ({ type, data, status: 'PAGE_NEXT', id }),
  ok: (type, data, id) => ({ type, data, status: 'OK', id }),
  nok: (type, data, id) => ({ type, data, status: 'NOK', id }),
  cancel: type => ({ type, status: 'CANCELLED' })
};

export {
  IssueFilter,
  notify,
  login
};

export default request;
