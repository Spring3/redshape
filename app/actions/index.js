import userActions from './user.actions';
import trackingActions from './tracking.actions';
import issuesActions from './issues.actions';
import projectActions from './project.actions';
import timeEntryActions from './timeEntry.actions';
import settingsActions from './settings.actions';

export default {
  user: userActions,
  issues: issuesActions,
  tracking: trackingActions,
  projects: projectActions,
  timeEntry: timeEntryActions,
  settings: settingsActions
};
