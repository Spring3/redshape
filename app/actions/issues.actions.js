import moment from 'moment';
import request, { notify } from './helper';

export const ISSUES_GET_ALL = 'ISSUES_GET_ALL';
export const ISSUES_GET = 'ISSUES_GET';
export const ISSUES_UPDATE = 'ISSUES_UPDATE';
export const ISSUES_COMMENT_SEND = 'ISSUES_COMMENT_SEND';
export const ISSUES_STATUSES_GET_ALL = 'ISSUES_STATUSES_GET_ALL';
export const ISSUES_TRACKERS_GET_ALL = 'ISSUES_TRACKERS_GET_ALL';

const getAll = (filter, offset, limit) => (dispatch) => {
  let url = '/issues.json?include=attachments,children,relations,journals';

  if (filter) {
    url += filter;
  }

  if (offset) {
    url += `&offset=${offset}`;
  }

  if (limit) {
    url += `&limit=${limit}`;
  }

  dispatch(notify.start(ISSUES_GET_ALL));

  return request({ url, id: 'getAllIssues' })
    .then(({ data }) => dispatch(notify.ok(ISSUES_GET_ALL, data)))
    .catch((error) => {
      console.error('Error when trying to get a list of issues:', error.message);
      dispatch(notify.nok(ISSUES_GET_ALL, error));
    });
};

const get = id => (dispatch) => {
  dispatch(notify.start(ISSUES_GET));

  return request({
    url: `/issues/${id}.json?include=attachments,children,relations,journals`,
    id: `getIssueDetails:${id}`
  }).then(({ data }) => dispatch(notify.ok(ISSUES_GET, data)))
    .catch((error) => {
      console.error(`Error when trying to get the issue with id ${id}:`, error.message);
      dispatch(notify.nok(ISSUES_GET, error));
    });
};

const sendComments = (issueId, comments) => (dispatch, getState) => {
  const { user = {} } = getState();
  const actionId = 'comments';

  dispatch(notify.start(ISSUES_COMMENT_SEND, actionId));

  return request({
    url: `/issues/${issueId}.json`,
    data: {
      issue: {
        notes: comments
      }
    },
    method: 'PUT'
  }).then(() => dispatch(
    notify.ok(
      ISSUES_COMMENT_SEND,
      {
        created_on: moment().toLocaleString(),
        details: [],
        id: Date.now(),
        notes: comments,
        private_notes: false,
        user: {
          id: user.id,
          name: user.name
        }
      },
      actionId
    )
  ))
    .catch((error) => {
      console.error(`Error when trying to assign the issue with id ${issueId}:`, error.message);
      dispatch(notify.nok(ISSUES_COMMENT_SEND, error, actionId));
    });
};

const getAllStatuses = () => (dispatch) => {
  dispatch(notify.start(ISSUES_STATUSES_GET_ALL));

  return request({
    url: '/issue_statuses.json',
    id: 'getAllIssueStatuses'
  }).then(({ data }) => dispatch(notify.ok(ISSUES_STATUSES_GET_ALL, data)))
    .catch((error) => {
      console.error('Error when trying to get the list of issue statuses:', error.message);
      dispatch(notify.nok(ISSUES_STATUSES_GET_ALL, error));
    });
};

const getAllTrackers = (dispatch) => {
  dispatch(notify.start(ISSUES_TRACKERS_GET_ALL));

  return request({
    url: '/trackers.json',
    id: 'getAllIssueTrackers'
  }).then(({ data }) => dispatch(notify.ok(ISSUES_TRACKERS_GET_ALL, data)))
    .catch((error) => {
      console.error('Error when trying to get the list of issue trackers:', error.message);
      dispatch(notify.nok(ISSUES_TRACKERS_GET_ALL, error));
    });
};

export default {
  getAll,
  get,
  sendComments,
  getAllStatuses,
  getAllTrackers
};
