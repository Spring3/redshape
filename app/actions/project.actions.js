import request, { notify } from './helper';

export const PROJECT_GET_ALL = 'PROJECT_GET_ALL';

const getAll = () => async (dispatch) => {
  dispatch(notify.start(PROJECT_GET_ALL));

  const requestPage = (offset = 0) => request({
    url: '/projects.json',
    query: {
      include: 'time_entry_activities',
      offset
    },
    id: `getProjects:${offset}`
  }).then(({ data }) => data);

  return requestPage()
    .then((data) => {
      const { total_count, offset, limit } = data;
      if (total_count > (offset + limit)) {
        // Math.floor because we already fetched the 1st page.
        const remainingPages = Math.floor(total_count / limit);
        const promises = [data];
        for (let i = 0; i < remainingPages; i++) {
          promises.push(requestPage((i + 1) * limit));
        }
        return Promise.all(promises);
      }
      return [data];
    })
    .then((results) => {
      dispatch(notify.ok(PROJECT_GET_ALL, results));
    })
    .catch((error) => {
      console.error('Error when trying to get the info about the existing projects', error);
      dispatch(notify.nok(PROJECT_GET_ALL, error));
    });
};

export default {
  getAll
};
