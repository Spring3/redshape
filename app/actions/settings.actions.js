export const SETTINGS_SHOW_CLOSED_ISSUES = 'SETTINGS_SHOW_CLOSED_ISSUES';
export const SETTINGS_USE_COLORS = 'SETTINGS_USE_COLORS';
export const SETTINGS_ISSUE_HEADERS = 'SETTINGS_ISSUE_HEADERS';
export const SETTINGS_BACKUP = 'SETTINGS_BACKUP';
export const SETTINGS_RESTORE = 'SETTINGS_RESTORE';

const setShowClosedIssues = showClosed => (dispatch, getState) => {
  const { user } = getState();
  dispatch({
    type: SETTINGS_SHOW_CLOSED_ISSUES,
    data: {
      userId: user.id,
      redmineEndpoint: user.redmineEndpoint,
      showClosed
    }
  });
};
const setUseColors = useColors => (dispatch, getState) => {
  const { user } = getState();
  dispatch({
    type: SETTINGS_USE_COLORS,
    data: {
      userId: user.id,
      redmineEndpoint: user.redmineEndpoint,
      useColors
    }
  });
};
const setIssueHeaders = data => (dispatch, getState) => {
  const { user } = getState();
  dispatch({
    type: SETTINGS_ISSUE_HEADERS,
    data: {
      userId: user.id,
      redmineEndpoint: user.redmineEndpoint,
      issueHeaders: data
    }
  });
};
const backup = () => (dispatch, getState) => {
  const { user } = getState();
  dispatch({
    type: SETTINGS_BACKUP,
    data: {
      userId: user.id,
      redmineEndpoint: user.redmineEndpoint
    }
  });
};
const restore = () => (dispatch, getState) => {
  const { user } = getState();
  dispatch({
    type: SETTINGS_RESTORE,
    data: {
      userId: user.id,
      redmineEndpoint: user.redmineEndpoint
    }
  });
};

export default {
  setShowClosedIssues,
  setUseColors,
  setIssueHeaders,
  backup,
  restore
};
