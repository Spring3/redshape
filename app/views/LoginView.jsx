import React, { Component } from 'react';
import _ from 'lodash';
import styled from 'styled-components';
import { Formik } from 'formik';
import Joi from 'joi';
import { withRouter } from 'react-router-dom';
import GithubCircleIcon from 'mdi-react/GithubCircleIcon';

import RedmineAPI from '../redmine/api.js';
import storage from '../../common/storage';

import { Input, Labeled } from '../components/Input';
import Button from '../components/Button';
import ErrorMessage from '../components/ErrorMessage';
import Link from '../components/Link';
import Copyrights from '../components/Copyrights';

const Container = styled.div`
  display: grid;
  grid-template-rows: repeat(4, 25vh);
  grid-template-columns: repeat(6, minmax(100px, 1fr));
  align-items: center;
  justify-items: center;
`;

const LoginForm = styled.form`
  padding: 40px;
  grid-column: 2 / 6;
  grid-row: 2 / 4;
  min-width: 300px;
`;

const Headline = styled.h1`
  text-align: center;
  font-size: 40px;
  color: #FF7079;
`;

const GHLinkContainer = styled.div`
  grid-column: 6;
  justify-self: end;
  align-self: start;
  margin: 20px 20px 0px 0px;
`;

const CopyrightsContainer = styled.div`
  grid-row: 4;
  grid-column: 2 / 6;
  align-self: end;
  margin-bottom: 20px;
`;

class LoginView extends Component {
  constructor(props) {
    super(props);
    const userData = storage.get('user');
    if (userData) {
      RedmineAPI.initialize(userData.redmineDomain).login({ token: userData.api_key });
      props.history.push('/app');
    }
  }

  validate = ({ username, password, redmineDomain }) => {
    const errors = {
      username: Joi.validate(username, Joi.string().required()),
      password: Joi.validate(password, Joi.string().required()),
      redmineDomain: Joi.validate(redmineDomain, Joi.string().uri().required())
    };
    const results = {};
    for (const [prop, validation] of Object.entries(errors)) {
      if (validation.error) {
        results[prop] = validation.error.message;
      }
    }
    return results;
  };

  onSubmit = (values, { setSubmitting, setFieldError }) => {
    const api = RedmineAPI.initialize(values.redmineDomain);
    api.login(values).then(({ data, error }) => {
      if (error) {
        setFieldError('request', error.message);
        
      } else {
        const user = _.get(data, 'user');
        if (user) {
          storage.set('user', {
            redmineDomain: values.redmineDomain,
            ..._.pick(user, 'id', 'firstname', 'lastname', 'api_key')
          });
          this.props.history.push('/app');
        }
      }
      setSubmitting(false);
    });
  }

  render() {
    return (
      <Container>
        <GHLinkContainer>
          <Link type="external" href="https://github.com/Spring3/redtime">
            <GithubCircleIcon color="#FF7079" size="30"/>
          </Link>
        </GHLinkContainer>
        <Formik
          initialValues={{ username: '', password: '', redmineDomain: '' }}
          validate={this.validate}
          onSubmit={this.onSubmit}
        >
          {({
            values,
            errors,
            touched,
            handleChange,
            handleBlur,
            handleSubmit,
            isSubmitting
          }) => (
            <LoginForm onSubmit={handleSubmit}>
              <Headline>Redtime</Headline>
              <Labeled
                label="Login"
                htmlFor="username"
              >
                <Input
                  type="text"
                  name="username"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.username}
                />
              </Labeled>
              <ErrorMessage show={errors.username && touched.username}>
                {errors.username}
              </ErrorMessage>
              <Labeled
                label="Password"
                htmlFor="password"
              >
                <Input
                  type="password"
                  name="password"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.password}
                />
              </Labeled>
              <ErrorMessage show={errors.password && touched.password}>
                {errors.password}
              </ErrorMessage>
              <Labeled
                label="Remdine Host"
                htmlFor="redmineDomain"
              >
                <Input
                  name="redmineDomain"
                  placeholder="https://redmine.example.com"
                  onChange={handleChange}
                  onBlur={handleBlur}
                  value={values.host}
                />
              </Labeled>
              <ErrorMessage show={errors.redmineDomain && touched.redmineDomain}>
                {errors.redmineDomain}
              </ErrorMessage>
              <Button
                type="submit"
                disabled={isSubmitting}
                block={true}
              >
                Submit
              </Button>
              <ErrorMessage show={errors.request}>
                {errors.request}
              </ErrorMessage>
            </LoginForm>
          )}
        </Formik>
        <CopyrightsContainer>
          <Copyrights />
        </CopyrightsContainer>
      </Container>
    );
  }
}

export default withRouter(LoginView);
