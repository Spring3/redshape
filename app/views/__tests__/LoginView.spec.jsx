import React from 'react';
import { HashRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { render, fireEvent, cleanup, wait } from 'react-testing-library';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';

import actions from '../../actions';
import { USER_LOGIN } from '../../actions/user.actions';
import { notify } from '../../actions/helper';

jest.mock('electron-store');

import storage from '../../../modules/storage';
import LoginView from '../LoginView';

const mockStore = configureStore([thunk]);

describe('Login view', () => {
  afterEach(() => {
    cleanup();
    fetch.resetMocks();
    storage.clear();
  });

  it('should render the link to github', () => {
    const store = mockStore({ user: {} });
    render(
      <Provider store={store}>
        <HashRouter><LoginView /></HashRouter>
      </Provider>
    );
    const link = document.querySelector('a[href="https://github.com/Spring3/redtime"]');
    expect(link).toBeTruthy();
    expect(store.getActions().length).toBe(0);

    store.clearActions();
  });

  it('should render the form', () => {
    const store = mockStore({ user: {} });
    render(
      <Provider store={store}>
        <HashRouter><LoginView /></HashRouter>
      </Provider>
    );
    const form = document.querySelector('form');
    expect(form).toBeTruthy();

    const usernameInput = form.querySelector('input[name="username"]');
    expect(usernameInput).toBeTruthy();
    expect(usernameInput.getAttribute('type')).toBe('text');

    const passwordInput = form.querySelector('input[type="password"]');
    expect(passwordInput).toBeTruthy();
    expect(passwordInput.getAttribute('name')).toBe('password');

    const redmineEndpointInput = form.querySelector('input[name="redmineEndpoint"]');
    expect(redmineEndpointInput).toBeTruthy();
    expect(redmineEndpointInput.getAttribute('type')).toBe('text');

    const submitButton = form.querySelector('button[type="submit"]');
    expect(submitButton).toBeTruthy();
    expect(submitButton.innerHTML).toBe('Submit');
    expect(store.getActions().length).toBe(0);

    store.clearActions();
  });

  it('should render the copyrights', () => {
    const store = mockStore({ user: {} });
    const { getByText } = render(
      <Provider store={store}>
        <HashRouter><LoginView /></HashRouter>
      </Provider>
    );
    const copyrights = getByText('Created by');
    expect(copyrights).toBeTruthy();
    const link = copyrights.querySelector('a');
    expect(link).toBeTruthy();
    expect(link.getAttribute('href')).toBeTruthy();
    expect(link.innerHTML).toBe('Daniyil Vasylenko');
    expect(store.getActions().length).toBe(0);

    store.clearActions();
  });

  it('should validate the username', () => {
    const store = mockStore({ user: {} });
    const { getByText } = render(
      <Provider store={store}>
        <HashRouter><LoginView /></HashRouter>
      </Provider>
    );
    const usernameInput = document.querySelector('input[name="username"]');
    expect(usernameInput).toBeTruthy();
    fireEvent.focus(usernameInput);
    fireEvent.blur(usernameInput);
    let error;
    return wait(() => {
      error = getByText('"username" is not allowed to be empty');
      expect(error).toBeTruthy();
      fireEvent.change(usernameInput, { target: { value: 'username' } });
    }).then(() => wait(() => {
      expect(error.innerHTML).toBeFalsy();
      expect(store.getActions().length).toBe(0);

      store.clearActions();
    }));
  });

  it('should validate the password', () => {
    const store = mockStore({ user: {} });
    const { getByText } = render(
      <Provider store={store}>
        <HashRouter><LoginView /></HashRouter>
      </Provider>
    );
    const passwordInput = document.querySelector('input[type="password"]');
    expect(passwordInput).toBeTruthy();
    fireEvent.focus(passwordInput);
    fireEvent.blur(passwordInput);
    let error;
    return wait(() => {
      error = getByText('"password" is not allowed to be empty');
      expect(error).toBeTruthy();
      fireEvent.change(passwordInput, { target: { value: 'password' } });
    }).then(() => wait(() => {
      expect(error.innerHTML).toBeFalsy();
      expect(store.getActions().length).toBe(0);

      store.clearActions();
    }));
  });

  it('should validate the redmine domain input', () => {
    const store = mockStore({ user: {} });
    const { getByText } = render(
      <Provider store={store}>
        <HashRouter><LoginView /></HashRouter>
      </Provider>
    );
    const redmineInput = document.querySelector('input[name="redmineEndpoint"]');
    expect(redmineInput).toBeTruthy();
    fireEvent.focus(redmineInput);
    fireEvent.blur(redmineInput);
    let error;
    return wait(() => {
      error = getByText('"redmineEndpoint" is not allowed to be empty');
      expect(error).toBeTruthy();
      fireEvent.change(redmineInput, { target: { value: 'url' } });
    }).then(() => wait(() => {
      error = getByText('"redmineEndpoint" must be a valid uri');
      expect(error).toBeTruthy();
      fireEvent.change(redmineInput, { target: { value: 'https://redmine.domain' } });
    }).then(() => wait(() => {
      expect(error.innerHTML).toBeFalsy();
      expect(store.getActions().length).toBe(0);

      store.clearActions();
    })));
  });

  it('should redirect to the AppView if user credentials already exist', async () => {
    const storeData = {
      user: {
        id: 1,
        firstname: 'firstname',
        lastname: 'lastname',
        api_key: 'api_key',
        redmineEndpoint: 'https://redmine.domain'
      }
    };
    const store = mockStore(storeData);
    const historyMock = {
      location: {
        pathname: '/'
      },
      listen: () => () => {},
      push: jest.fn()
    };

    render(
      <Provider store={store}>
        <Router history={historyMock}><LoginView /></Router>
      </Provider>
    );
    await wait(() => {
      expect(store.getActions().length).toBe(0);
      store.clearActions();
    });
  });
});

describe('[integration] LoginView', () => {
  afterEach(() => {
    fetch.resetMocks();
    storage.clear();
  });

  it('should not make a redmine api request if the form has errors', (done) => {
    const store = mockStore({ user: {} });
    const storageSetSpy = jest.spyOn(storage, 'set');
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter><LoginView /></HashRouter>
      </Provider>
    );
    const form = wrapper.find('LoginView');
    expect(form.exists()).toBeTruthy();
    const submitButton = wrapper.find('button[type="submit"]');
    expect(submitButton.exists()).toBeTruthy();
    submitButton.simulate('submit');
    setTimeout(() => {
      const error = wrapper.find('ErrorMessage').findWhere(item => /is not allowed to be empty/.test(item.text()));
      expect(error.exists()).toBeTruthy();
      expect(storageSetSpy).not.toHaveBeenCalled();
      expect(store.getActions().length).toBe(0);

      storageSetSpy.mockRestore();
      store.clearActions();
      done();
    }, 100);
  });

  it('should make a redmine api request on submit', (done) => {
    const store = mockStore({ user: {} });
    const loginActionSpy = jest.spyOn(actions.user, 'checkLogin');
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter><LoginView /></HashRouter>
      </Provider>
    );
    const form = wrapper.find('LoginView');
    expect(form.exists()).toBeTruthy();

    const userData = {
      id: '123',
      firstname: 'test',
      lastname: 'user',
      api_key: 'api_key'
    };
    fetch.mockResponseOnce(JSON.stringify({ user: userData }));

    const submit = wrapper.find('button[type="submit"]');
    expect(submit.exists()).toBeTruthy();

    const inputs = wrapper.find('input');
    expect(inputs.length).toBe(3);
    const usernameInput = inputs.at(0);
    const passwordInput = inputs.at(1);
    const redmineEndpointInput = inputs.at(2);
    expect(usernameInput.prop('name')).toBe('username');
    expect(passwordInput.prop('name')).toBe('password');
    expect(redmineEndpointInput.prop('name')).toBe('redmineEndpoint');

    const returnedValues = {
      username: 'username',
      password: 'password',
      redmineEndpoint: 'https://redmine.domain'
    };

    usernameInput.simulate('change', { persist: () => {}, target: { value: returnedValues.username, name: 'username' } });
    passwordInput.simulate('change', { persist: () => {}, target: { value: returnedValues.password, name: 'password' } });
    redmineEndpointInput.simulate('change', { persist: () => {}, target: { name: 'redmineEndpoint', value: returnedValues.redmineEndpoint } });
    expect(wrapper.find('input[name="username"]').prop('value')).toBe(returnedValues.username);
    expect(wrapper.find('input[name="password"]').prop('value')).toBe(returnedValues.password);
    expect(wrapper.find('input[name="redmineEndpoint"]').prop('value')).toBe(returnedValues.redmineEndpoint);

    submit.simulate('submit');

    setTimeout(() => {
      wrapper.update();
      const errors = wrapper.find('ErrorMessage').findWhere(item => /is not allowed to be empty/.test(item.text()));
      expect(errors.exists()).toBeFalsy();

      expect(fetch.mock.calls.length).toBe(1);
      expect(fetch.mock.calls[0][0]).toBe(`${returnedValues.redmineEndpoint}/users/current.json`);
      expect(fetch.mock.calls[0][1].headers.Authorization).toBe(`Basic ${btoa(`${returnedValues.username}:${returnedValues.password}`)}`);
      expect(fetch.mock.calls[0][1].method).toBe('GET');

      expect(loginActionSpy).toHaveBeenCalledWith(returnedValues);
      const reduxActions = store.getActions();
      expect(reduxActions.length).toBe(2);
      expect(reduxActions[0]).toEqual(notify.start(USER_LOGIN));
      expect(reduxActions[1]).toEqual(notify.ok(USER_LOGIN, {
        user: {
          ...userData,
          redmineEndpoint: returnedValues.redmineEndpoint
        }
      }));

      store.clearActions();
      loginActionSpy.mockRestore();
      done();
    }, 100);
  });

  it('should display the error if one raised during the request', async () => {
    // simulating error since reducers are not called
    const store = mockStore({ user: { loginError: new Error('Something went wrong') } });
    const loginActionSpy = jest.spyOn(actions.user, 'checkLogin');
    const wrapper = mount(
      <Provider store={store}>
        <HashRouter><LoginView /></HashRouter>
      </Provider>
    );
    const form = wrapper.find('LoginView');
    expect(form.exists()).toBeTruthy();

    const expectedError = new Error('Something went wrong');

    fetch.mockReject(expectedError);

    const inputs = wrapper.find('input');
    expect(inputs.length).toBe(3);
    const usernameInput = inputs.at(0);
    const passwordInput = inputs.at(1);
    const redmineEndpointInput = inputs.at(2);
    expect(usernameInput.prop('name')).toBe('username');
    expect(passwordInput.prop('name')).toBe('password');
    expect(redmineEndpointInput.prop('name')).toBe('redmineEndpoint');

    const submitButton = wrapper.find('button[type="submit"]');
    expect(submitButton.exists()).toBeTruthy();

    const returnedValues = {
      username: 'username',
      password: 'password',
      redmineEndpoint: 'https://redmine.domain'
    };

    usernameInput.simulate('change', { persist: () => {}, target: { value: returnedValues.username, name: 'username' } });
    passwordInput.simulate('change', { persist: () => {}, target: { value: returnedValues.password, name: 'password' } });
    redmineEndpointInput.simulate('change', { persist: () => {}, target: { name: 'redmineEndpoint', value: returnedValues.redmineEndpoint } });
    expect(wrapper.find('input[name="username"]').prop('value')).toBe(returnedValues.username);
    expect(wrapper.find('input[name="password"]').prop('value')).toBe(returnedValues.password);
    expect(wrapper.find('input[name="redmineEndpoint"]').prop('value')).toBe(returnedValues.redmineEndpoint);

    submitButton.simulate('submit');

    await wait(() => {
      wrapper.update();
      expect(loginActionSpy).toHaveBeenCalledWith(returnedValues);
      expect(fetch.mock.calls.length).toBe(1);
      expect(fetch.mock.calls[0][0]).toBe(`${returnedValues.redmineEndpoint}/users/current.json`);
      expect(fetch.mock.calls[0][1].headers.Authorization).toBe(`Basic ${btoa(`${returnedValues.username}:${returnedValues.password}`)}`);
      expect(fetch.mock.calls[0][1].method).toBe('GET');

      const error = wrapper.find('ErrorMessage').findWhere(item => /Something went wrong/.test(item.text()));
      expect(error.exists()).toBeTruthy();
      const reduxActions = store.getActions();
      expect(reduxActions.length).toBe(2);
      expect(reduxActions[0]).toEqual(notify.start(USER_LOGIN));
      expect(reduxActions[1]).toEqual(notify.nok(USER_LOGIN, expectedError));

      store.clearActions();
      loginActionSpy.mockRestore();
    });
  });
});
