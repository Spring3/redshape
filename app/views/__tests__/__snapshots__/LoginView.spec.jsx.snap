// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login view should match the snapshot 1`] = `
.c6 {
  display: block;
  width: 100%;
  border-radius: 5px;
  padding: 5px 10px;
  box-sizing: border-box;
  font-size: 14px;
  border: 1px solid #FF7079;
  min-height: 35px;
  outline: none;
  color: #FF7079;
  font-weight: bold;
  background: #FAFAFA;
}

.c6:hover {
  border: 1px solid #FF7079;
}

.c6:focus {
  box-shadow: 0px 0px 0px 1px #FF7079;
}

.c5 h4 {
  margin-bottom: 10px;
  color: #A4A4A4;
}

.c8 {
  padding: 10px 5px;
  border-radius: 5px;
  border: 2px solid transparent;
  box-shadow: 0px 0px 0px 2px #FF7079;
  color: #FF7079;
  font-weight: bold;
  font-size: 14px;
  outline: none;
  width: 100%;
  margin: 25px auto 0px auto;
}

.c8:hover,
.c8:focus {
  cursor: pointer;
  background: #FF7079;
  border: 2px solid #FF7079;
  box-shadow: 0px 0px 0px 2px #EF6069;
  color: white;
}

.c8:active {
  background: #EF6069;
}

.c7 {
  margin: 15px auto 5px auto;
  font-size: 14px;
  background: #FF7079;
  color: white;
  padding: 2px;
  display: none;
}

.c2 {
  color: #FF7079;
  font-size: 14px;
  padding: 2px;
}

.c11 {
  color: #FF7079;
  font-size: 14px;
  padding: 2px;
}

.c11:hover {
  background: #FF7079;
  color: white;
}

.c10 {
  color: #FF7079;
}

.c0 {
  display: grid;
  grid-template-rows: repeat(4,25vh);
  grid-template-columns: repeat(6,minmax(100px,1fr));
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
}

.c3 {
  padding: 40px;
  grid-column: 2 / 6;
  grid-row: 2 / 4;
  min-width: 300px;
}

.c4 {
  text-align: center;
  font-size: 40px;
  color: #FF7079;
}

.c1 {
  grid-column: 6;
  justify-self: end;
  -webkit-align-self: start;
  -ms-flex-item-align: start;
  align-self: start;
  margin: 20px 20px 0px 0px;
}

.c9 {
  grid-row: 4;
  grid-column: 2 / 6;
  -webkit-align-self: end;
  -ms-flex-item-align: end;
  align-self: end;
  margin-bottom: 20px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <a
      className="c2"
      href="https://github.com/Spring3/redtime"
      rel="noopener noreferer"
      type="external"
    >
      <svg
        className="mdi-icon "
        fill="#FF7079"
        height="30"
        viewBox="0 0 24 24"
        width="30"
      >
        <path
          d="M12,2C6.48,2 2,6.48 2,12C2,16.42 4.87,20.17 8.84,21.5C9.34,21.58 9.5,21.27 9.5,21C9.5,20.77 9.5,20.14 9.5,19.31C6.73,19.91 6.14,17.97 6.14,17.97C5.68,16.81 5.03,16.5 5.03,16.5C4.12,15.88 5.1,15.9 5.1,15.9C6.1,15.97 6.63,16.93 6.63,16.93C7.5,18.45 8.97,18 9.54,17.76C9.63,17.11 9.89,16.67 10.17,16.42C7.95,16.17 5.62,15.31 5.62,11.5C5.62,10.39 6,9.5 6.65,8.79C6.55,8.54 6.2,7.5 6.75,6.15C6.75,6.15 7.59,5.88 9.5,7.17C10.29,6.95 11.15,6.84 12,6.84C12.85,6.84 13.71,6.95 14.5,7.17C16.41,5.88 17.25,6.15 17.25,6.15C17.8,7.5 17.45,8.54 17.35,8.79C18,9.5 18.38,10.39 18.38,11.5C18.38,15.32 16.04,16.16 13.81,16.41C14.17,16.72 14.5,17.33 14.5,18.26C14.5,19.6 14.5,20.68 14.5,21C14.5,21.27 14.66,21.59 15.17,21.5C19.14,20.16 22,16.42 22,12C22,6.48 17.52,2 12,2Z"
        />
      </svg>
    </a>
  </div>
  <form
    className="c3"
    onSubmit={[Function]}
  >
    <h1
      className="c4"
    >
      Redtime
    </h1>
    <div
      className="form-group undefined c5"
    >
      <h4
        htmlFor="username"
      >
        Login
      </h4>
      <input
        className="c6"
        disabled={false}
        name="username"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <p
      className="c7"
    />
    <div
      className="form-group undefined c5"
    >
      <h4
        htmlFor="password"
      >
        Password
      </h4>
      <input
        className="c6"
        disabled={false}
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <p
      className="c7"
    />
    <div
      className="form-group undefined c5"
    >
      <h4
        htmlFor="redmineEndpoint"
      >
        Remdine Host
      </h4>
      <input
        className="c6"
        disabled={false}
        name="redmineEndpoint"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="https://redmine.example.com"
        type="text"
        value=""
      />
    </div>
    <p
      className="c7"
    />
    <button
      className="c8"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
    <p
      className="c7"
    />
  </form>
  <div
    className="c9"
  >
    <p
      className="c10"
    >
      Created byÂ 
      <a
        className="c11"
        href="https://github.com/Spring3"
        rel="noopener noreferer"
        type="external"
      >
        Daniyil Vasylenko
      </a>
    </p>
  </div>
</div>
`;
